src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -3131,13 +3131,44 @@ trait Types
        */
       def unifyFull(tpe: Type): Boolean = {
         def unifySpecific(tp: Type) = {
-          sameLength(typeArgs, tp.typeArgs) && {
+          if(sameLength(typeArgs, tp.typeArgs)) {
             val lhs = if (isLowerBound) tp.typeArgs else typeArgs
             val rhs = if (isLowerBound) typeArgs else tp.typeArgs
             // This is a higher-kinded type var with same arity as tp.
             // If so (see SI-7517), side effect: adds the type constructor itself as a bound.
             isSubArgs(lhs, rhs, params, AnyDepth) && { addBound(tp.typeConstructor); true }
-          }
+          } else if(settings.YhigherOrderUnification && typeArgs.lengthCompare(0) > 0 && compareLengths(typeArgs, tp.typeArgs) < 0) {
+            // Simple algorithm as suggested by Paul Chiusano in the comments on SI-2712
+            //
+            //   https://issues.scala-lang.org/browse/SI-2712?focusedCommentId=61270
+            //
+            // Treat the type constructor as curried and partially applied, we treat a prefix
+            // as constants and solve for the suffix. For the example in the ticket, unifying
+            // M[A] with Int => Int this unifies as,
+            //
+            //   M[t] = [t][Int => t]
+            //   A = Int
+            //
+            // A more "natural" unifier might be M[t] = [t][t => t]. There's lots of scope for
+            // experimenting with alternatives here.
+
+            val captured = tp.typeArgs.length-typeArgs.length
+            val (prefix, suffix) = tp.typeArgs.splitAt(captured)
+
+            val lhs = if (isLowerBound) suffix else typeArgs
+            val rhs = if (isLowerBound) typeArgs else suffix
+            // This is a higher-kinded type var with same arity as tp.
+            // If so (see SI-7517), side effect: adds the type constructor itself as a bound.
+            isSubArgs(lhs, rhs, params, AnyDepth) && {
+              val tpSym = tp.typeSymbolDirect
```

Comment:
This is lossy: the prefix of the `tp` will not contribute to the bounds of the synthetic type params.

```
scala> class C[A] { trait Two[_1 <: A, _2 <: A]; type One[_1 <: A] = Two[_1, _1] }
defined class C

scala> def test(cint: C[Int])(two: cint.Two[Int, Int]) = f(two)
<console>:13: error: inferred kinds of the type arguments ([_2 <: A]cint.Two[Int,_2]) do not conform to the expected kinds of the type parameters (type M).
[_2 <: A]cint.Two[Int,_2]'s type parameters do not match type M's expected parameters:
type _2's bounds <: A are stricter than type _'s declared bounds <: Int
       def test(cint: C[Int])(two: cint.Two[Int, Int]) = f(two)
                                                         ^
```

That said, it appears we have the same problem with regular inference, so we can open a ticket to followup and fix both cases:

```
scala> def test(cint: C[Int])(two: cint.Two[Int, Int]) = { f[cint.One](two) }
<console>:13: error: kinds of the type arguments (cint.One) do not conform to the expected kinds of the type parameters (type M).
cint.One's type parameters do not match type M's expected parameters:
type _1's bounds <: A are stricter than type _'s declared bounds <: Int
       def test(cint: C[Int])(two: cint.Two[Int, Int]) = { f[cint.One](two) }
                                                            ^

scala> def test(cint: C[Int])(two: cint.Two[Int, Int]) = { type One1[_1 <: Int] = cint.One[_1]; f[One1](two) }
test: (cint: C[Int])(two: cint.Two[Int,Int])One1[Int]
```